### DISCUSSION TITLE
Window rules to center floating window on another window

### DISCUSSION THREAD

**Discussion by omnigenous** [hyprwm/Hyprland#7144] [Q&A]
I know `class` and `title` of parent window that I want to center child floating window on. I couldn't figure out if this possible after reading https://wiki.hyprland.org/Configuring/Window-Rules/ but I don't fully understand `move` section.

So far I got:
```
windowrulev2 = float, class:App title:App child
windowrulev2 = size 900 500, class:App title:App child
# now center it on parent window
```

I know `center` exists but I don't want to center on _monitor_ I want to center on specific parent window with same class.

**Comment by DHDcc** [comment]
@vaxerski

**Comment by vaxerski** [comment]
I dont think this is possible as we dont do math in move rules.

**Comment by Hannes1909** [comment]
Would this be possible with LUA? O:-)
Most likely rhetorical question, I'm sure...

**Reply by sagayda** [reply]
I doubt this is achievable with window rules, but can be done with events:
```lua
hl.on("window.open", function(window)
	if window.class == "child" then
		local parent_window = hl.get_window("class:^(parent)$")
		if parent_window then
			local parent_at = parent_window.at
			local parent_size = parent_window.size
			local child_size = window.size

			local x = (parent_size.x - child_size.x) / 2 + parent_at.x
			local y = (parent_size.y - child_size.y) / 2 + parent_at.y

			hl.dispatch(hl.dsp.window.move({ x = x, y = y, window = window }))
		end
	end
end)
```
Fun fact - using this with `window.open_early` crashed my hyprland, so use it only with `window.open`)))

UPD: also works with `window.update_rules`, which might be even better if the child window changes it's title\class

**Reply by omnigenous** [reply]
@sagayda Can I match child window using title Regular expression as well? I want to center particular child window that has title like this: "Setting - APPNAME v2.4.5"

**Reply by sagayda** [reply]
> @sagayda Can I match child window using title Regular expression as well? I want to center particular child window that has title like this: "Setting - APPNAME v2.4.5"

I couldn't find any hyprland lua api functions that expose its regex. However, you can still use lua's built-in [pattern matching](https://www.lua.org/pil/20.1.html):
```lua
window.title:match("^Settings %- .- v%d+%.%d+%.%d+$")
```
will match any name or version.

Or query the child window with `hl.get_window` and compare the objects:
```lua
local target_window = hl.get_window("title:^(regex)$")
if window == target_window then
   -- ...
end
```

**Reply by omnigenous** [reply]
I will try that. But can I "get_window" with both "title" and "class"? I can't get it to work.

**Reply by sagayda** [reply]
Can't say for sure, but most likely, no - https://wiki.hypr.land/Configuring/Basics/Dispatchers/#window

**Reply by ItsOhen** [reply]
No. No way for hyprland to know if a window is from a parent window or not.

**Reply by omnigenous** [reply]
Okay, I need to experiment more on my own. So far I wrapped your code into function and now I'm not sure if it is a good idea to call this function say 50 times in different places. Would this noticeably degrade performance? I don't have 50 apps to actually test this.

```
class_var=""
my_func(class_var)
...
```

**Reply by omnigenous** [reply]
@ItsOhen You are replying to @Hannes1909  ?

**Reply by sagayda** [reply]
> I don't have 50 apps to actually test this.

So why bother?)



> Would this noticeably degrade performance?

```lua
local startTime = os.clock()

for i = 1, 1000, 1 do
    local target_window = hl.get_window("class:^(app)$")
    if window == target_window then
        print("1")
    end
end

local endTime = os.clock()
local elapsedTime = endTime - startTime
```
0.058-0.06 seconds for 1000 checks, so roughly 0.06ms per check on my system
